tests are sensitive to docker buildkit configuration

### Bug description

Our docker tests are sensitive to docker build output, but that output is different depending on whether buildkit is enabled or not (reported in #190). The build output may even be different according to localization (I don't know if docker terminal output is localized).

The reason is that we have checks for text output like:

```
Successfully tagged testchart/testimage:0.0.1-0.dev.git.1.h86465ea
```

but the same event looks like:

```
naming to docker.io/testchart/testimage:0.0.1-0.dev.git.1.h86465ea done
```

when run with buildkit.

#### Expected behaviour

Tests pass whether buildkit is enabled or not, or tests enforce the expected configuration.

#### Actual behaviour

Tests fail when buildkit is enabled with e.g.

<details>
<summary>

```
FAILED tests/test_repo_interactions.py::test_chartpress_run[None] - assert 'Successfully tagged testchart/testimage:0.0.1-0.dev.git.1.h86465ea' in "...
```

</summary>

```
$ pytest --ff -m 'not registry' -vx tests/
================================================================================================================================================================================================== test session starts ==================================================================================================================================================================================================
platform darwin -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- /Users/minrk/conda/bin/python
cachedir: .pytest_cache
rootdir: /Users/minrk, configfile: pytest.ini
plugins: anyio-3.6.1, xdist-2.5.0, forked-1.4.0, asyncio-0.19.0, timeout-2.1.0, cov-3.0.0
asyncio: mode=strict
collected 68 items / 1 deselected / 67 selected
run-last-failure: rerun previous 10 failures first

tests/test_repo_interactions.py::test_git_repo_fixture PASSED                                                                                                                                                                                                                                                                                                                                                     [  1%]
tests/test_repo_interactions.py::test_chartpress_run[None] FAILED                                                                                                                                                                                                                                                                                                                                                 [  2%]

======================================================================================================================================================================================================= FAILURES ========================================================================================================================================================================================================
_______________________________________________________________________________________________________________________________________________________________________________________________ test_chartpress_run[None] _______________________________________________________________________________________________________________________________________________________________________________________________

git_repo = <git.repo.base.Repo '/var/folders/qr/3vxfnp1x2t1fw55dr288mphc0000gn/T/tmpvaijnbno/.git'>, capfd = <_pytest.capture.CaptureFixture object at 0x1056e3c70>, base_version = None

    @pytest.mark.parametrize("base_version", [None, "0.0.1-0.dev"])
    def test_chartpress_run(git_repo, capfd, base_version):
        """Run chartpress and inspect the output."""

        with open("chartpress.yaml") as f:
            chartpress_config = yaml.load(f)

        chart = chartpress_config["charts"][0]
        if base_version:
            chart["baseVersion"] = base_version
            with open("chartpress.yaml", "w") as f:
                yaml.dump(chartpress_config, f)

        # summarize information from git_repo
        sha = git_repo.commit("HEAD").hexsha[:7]
        tag = f"0.0.1-{PRERELEASE_PREFIX}.1.h{sha}"
        check_version(tag)

        # run chartpress
        out = _capture_output([], capfd)
        print(out)
        # verify image was built
        # verify the fallback tag of "0.0.1" when a tag is missing
>       assert f"Successfully tagged testchart/testimage:{tag}" in out
E       assert 'Successfully tagged testchart/testimage:0.0.1-0.dev.git.1.h86465ea' in "Updating testchart/Chart.yaml: version: 0.0.1-0.dev.git.1.h86465ea\n$> docker build -t testchart/testimage:0.0.1-0.dev.git.1.h86465ea image -f image/Dockerfile --build-arg TEST_STATIC_BUILD_ARG=test --build-arg TEST_DYNAMIC_BUILD_ARG=0.0.1-0.dev.git.1.h86465ea-86465ea --label=maintainer=octocat --label ref=0.0.1-0.dev.git.1.h86465ea-86465ea --rm\n#1 [internal] load build definition from Dockerfile\n#1 sha256:c41d89e0acfd821ffe2bf6b34c64b4d44ab3f07d2f6594ceaa73eca2a1b255a5\n#1 transferring dockerfile: 37B done\n#1 DONE 0.0s\n\n#2 [internal] load .dockerignore\n#2 sha256:84299fcf8593eb7ca70e50059087d52aa87a443e1cd9cac2764c1e37b291cb2c\n#2 transferring context: 2B done\n#2 DONE 0.0s\n\n#3 [internal] load metadata for docker.io/library/alpine:latest\n#3 sha256:13549c58a76bcb5dac9d52bc368a8fb6b5cf7659f94e3fa6294917b85546978d\n#3 DONE 0.0s\n\n#4 [1/1] FROM docker.io/library/alpine\n#4 sha256:d20daa00e252bfb345a1b4f53b6bb332aafe702d8de5e583a76fcd09ba7ea1c1\n#4 CACHED\n\n#5 exporting to image\n#5 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00\n#5 exporting layers done\n#5 writing image sha256:a1af254e844927030e12a8d24a327f0731c7160a0a42ee6fb3167b9085bd10cd done\...4bfcbdecaf5\n#1 transferring dockerfile: 36B done\n#1 DONE 0.0s\n\n#2 [internal] load .dockerignore\n#2 sha256:ebaba897af5a6e791520e5799cef0e189200ffd326e24adf9bcbc3c737710135\n#2 transferring context: 2B done\n#2 DONE 0.0s\n\n#3 [internal] load metadata for docker.io/library/busybox:latest\n#3 sha256:0a79f84dd133133d07fb884b418e9bf0d7cf3d0c3447f64349c6cd669102adf5\n#3 DONE 0.0s\n\n#4 [1/1] FROM docker.io/library/busybox\n#4 sha256:95072eaeb63927804e292abcc951bc29637da310b8decf274190a5eaf5f97694\n#4 CACHED\n\n#5 exporting to image\n#5 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00\n#5 exporting layers done\n#5 writing image sha256:33fcdaa8edd1bccbc6310b8a4f6d65d6d3ec07a5ce47c452ef429296cd88d9a9 done\n#5 naming to docker.io/testchart/amd64only:0.0.1-0.dev.git.1.h86465ea done\n#5 DONE 0.0s\n\nUse 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them\nUpdating testchart/values.yaml: image: testchart/testimage:0.0.1-0.dev.git.1.h86465ea\nUpdating testchart/values.yaml: list.0: testchart/testimage:0.0.1-0.dev.git.1.h86465ea\nUpdating testchart/values.yaml: list.1.image: testchart/testimage:0.0.1-0.dev.git.1.h86465ea\n"

/Users/minrk/dev/jpy/hub/chartpress/tests/test_repo_interactions.py:55: AssertionError
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Updating testchart/Chart.yaml: version: 0.0.1-0.dev.git.1.h86465ea
$> docker build -t testchart/testimage:0.0.1-0.dev.git.1.h86465ea image -f image/Dockerfile --build-arg TEST_STATIC_BUILD_ARG=test --build-arg TEST_DYNAMIC_BUILD_ARG=0.0.1-0.dev.git.1.h86465ea-86465ea --label=maintainer=octocat --label ref=0.0.1-0.dev.git.1.h86465ea-86465ea --rm
#1 [internal] load build definition from Dockerfile
#1 sha256:c41d89e0acfd821ffe2bf6b34c64b4d44ab3f07d2f6594ceaa73eca2a1b255a5
#1 transferring dockerfile: 37B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:84299fcf8593eb7ca70e50059087d52aa87a443e1cd9cac2764c1e37b291cb2c
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/alpine:latest
#3 sha256:13549c58a76bcb5dac9d52bc368a8fb6b5cf7659f94e3fa6294917b85546978d
#3 DONE 0.0s

#4 [1/1] FROM docker.io/library/alpine
#4 sha256:d20daa00e252bfb345a1b4f53b6bb332aafe702d8de5e583a76fcd09ba7ea1c1
#4 CACHED

#5 exporting to image
#5 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
#5 exporting layers done
#5 writing image sha256:a1af254e844927030e12a8d24a327f0731c7160a0a42ee6fb3167b9085bd10cd done
#5 naming to docker.io/testchart/testimage:0.0.1-0.dev.git.1.h86465ea done
#5 DONE 0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
$> docker build -t testchart/amd64only:0.0.1-0.dev.git.1.h86465ea amd64only -f amd64only/Dockerfile
#1 [internal] load build definition from Dockerfile
#1 sha256:5fc5be3a4a38c27fa00b3a2038ea8b44a08b4d03dfde5eb1186334bfcbdecaf5
#1 transferring dockerfile: 36B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:ebaba897af5a6e791520e5799cef0e189200ffd326e24adf9bcbc3c737710135
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/busybox:latest
#3 sha256:0a79f84dd133133d07fb884b418e9bf0d7cf3d0c3447f64349c6cd669102adf5
#3 DONE 0.0s

#4 [1/1] FROM docker.io/library/busybox
#4 sha256:95072eaeb63927804e292abcc951bc29637da310b8decf274190a5eaf5f97694
#4 CACHED

#5 exporting to image
#5 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
#5 exporting layers done
#5 writing image sha256:33fcdaa8edd1bccbc6310b8a4f6d65d6d3ec07a5ce47c452ef429296cd88d9a9 done
#5 naming to docker.io/testchart/amd64only:0.0.1-0.dev.git.1.h86465ea done
#5 DONE 0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
Updating testchart/values.yaml: image: testchart/testimage:0.0.1-0.dev.git.1.h86465ea
Updating testchart/values.yaml: list.0: testchart/testimage:0.0.1-0.dev.git.1.h86465ea
Updating testchart/values.yaml: list.1.image: testchart/testimage:0.0.1-0.dev.git.1.h86465ea

=================================================================================================================================================================================================== warnings summary ====================================================================================================================================================================================================
dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
  /Users/minrk/conda/lib/python3.10/site-packages/docker/utils/utils.py:52: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    s1 = StrictVersion(v1)

dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
dev/jpy/hub/chartpress/tests/test_repo_interactions.py::test_chartpress_run[None]
  /Users/minrk/conda/lib/python3.10/site-packages/docker/utils/utils.py:53: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    s2 = StrictVersion(v2)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================================================================================================================ short test summary info ================================================================================================================================================================================================
FAILED tests/test_repo_interactions.py::test_chartpress_run[None] - assert 'Successfully tagged testchart/testimage:0.0.1-0.dev.git.1.h86465ea' in "Updating testchart/Chart.yaml: version: 0.0.1-0.dev.git.1.h86465ea\n$> docker build -t testchart/testimage:0.0.1-0.dev.git.1.h86465ea image -f image/Dockerfile --build-arg TEST_STATIC_BUILD_ARG=test --build-arg TEST_DYNAMIC_BUILD_ARG=0.0.1-0.dev.git.1.h86465...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================================================================================================= 1 failed, 1 passed, 1 deselected, 6 warnings in 5.31s =================================================================================================================================================================================
```

</details>

<!-- Tell us what actually happens. -->

### How to reproduce

<!-- Use this section to describe the steps that a user would take to experience this bug. -->

1. `export DOCKER_BUILDKIT=1` or enable buildkit in your docker config
2. `pytest -vx tests/test_repo_interactions.py`


### Proposed fixes

Two possible fixes (why just one, when we can do both!):

1. simplest short-term fix is to explicitly set `DOCKER_BUILDKIT=0` in the tests themselves (or `1` and update output expectations).
2. the more robust solution is to stop checking docker terminal output, and instead check the actual effects of the tests, e.g. via docker API calls (does this image:tag exist, etc.)

--- a/.github/workflows/publish.yaml
+++ b/.github/workflows/publish.yaml
@@ -1,5 +1,5 @@
 # This is a GitHub workflow defining a set of jobs with a set of steps.
-# ref: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions
+# ref: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
 #
 # Publish to PyPI on push of version like tags
 #
@@ -24,12 +24,12 @@ on:
 
 jobs:
   publish-to-pypi:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v4
         with:
-          python-version: "3.9"
+          python-version: "3.10"
 
       - name: install build package
         run: |

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -1,5 +1,5 @@
 # This is a GitHub workflow defining a set of jobs with a set of steps.
-# ref: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions
+# ref: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
 #
 name: Test
 
@@ -21,7 +21,7 @@ on:
 
 jobs:
   chart-tests:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
 
     strategy:
       # Keep running even if one variation of the job fail
@@ -36,6 +36,7 @@ jobs:
           - python: "3.8"
           - python: "3.9"
           - python: "3.10"
+          - python: "3.11"
 
     services:
       local-registry:

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -102,5 +102,5 @@ jobs:
         run: |
           pytest --verbose --color=yes --cov chartpress
 
-      - name: Upload coverage report
-        run: codecov
+      # GitHub action reference: https://github.com/codecov/codecov-action
+      - uses: codecov/codecov-action@v3

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -1,5 +1,4 @@
 bump2version
-codecov
 gitpython
 pre-commit
 pytest

--- a/tests/test_chartpress.py
+++ b/tests/test_chartpress.py
@@ -117,15 +117,15 @@ def test_buildx(git_repo, capfd):
     # Note: Parsing stderr for these logs may be fragile. We could probably omit
     # it and just verify the registry manifests instead.
 
-    assert "[linux/amd64 1/1] FROM docker.io/library/alpine@" in stderr
-    assert "[linux/arm64 1/1] FROM docker.io/library/alpine@" in stderr
-    assert "[linux/ppc64le 1/1] FROM docker.io/library/alpine@" in stderr
+    assert "[linux/amd64 1/1] FROM docker.io/library/alpine" in stderr
+    assert "[linux/arm64 1/1] FROM docker.io/library/alpine" in stderr
+    assert "[linux/ppc64le 1/1] FROM docker.io/library/alpine" in stderr
 
     # If there's only one platform it doesn't appear in the output logs
-    # assert "[linux/amd64 1/1] FROM docker.io/library/busybox@" in stderr
-    assert "[1/1] FROM docker.io/library/busybox@" in stderr
-    assert "[linux/arm64 1/1] FROM docker.io/library/busybox@" not in stderr
-    assert "[linux/ppc64le 1/1] FROM docker.io/library/busybox@" not in stderr
+    # assert "[linux/amd64 1/1] FROM docker.io/library/busybox" in stderr
+    assert "[1/1] FROM docker.io/library/busybox" in stderr
+    assert "[linux/arm64 1/1] FROM docker.io/library/busybox" not in stderr
+    assert "[linux/ppc64le 1/1] FROM docker.io/library/busybox" not in stderr
 
     assert f"pushing manifest for localhost:5000/test-buildx/testimage:{tag}" in stderr
     assert f"pushing manifest for localhost:5000/test-buildx/amd64only:{tag}" in stderr

--- a/tests/test_repo_interactions.py
+++ b/tests/test_repo_interactions.py
@@ -53,7 +53,10 @@ def test_chartpress_run(git_repo, capfd, base_version):
     print(out)
     # verify image was built
     # verify the fallback tag of "0.0.1" when a tag is missing
-    assert f"Successfully tagged testchart/testimage:{tag}" in out
+    assert (
+        f"Successfully tagged testchart/testimage:{tag}" in out
+        or f"naming to docker.io/testchart/testimage:{tag} in out"
+    )
 
     # verify the passing of static and dynamic --build-args
     assert "--build-arg TEST_STATIC_BUILD_ARG=test" in out
@@ -83,12 +86,12 @@ def test_chartpress_run(git_repo, capfd, base_version):
 
     # verify usage of --force-build
     out = _capture_output(["--force-build"], capfd)
-    assert "Successfully tagged" in out
+    assert "Successfully tagged" in out or "naming to" in out
 
     # verify usage --skip-build and --tag
     tag = "1.2.3-test.tag"
     out = _capture_output(["--skip-build", "--tag", tag], capfd)
-    assert "Successfully tagged" not in out
+    assert "Successfully tagged" not in out or "naming to" in out
     assert f"Updating testchart/Chart.yaml: version: {tag}" in out
     assert f"Updating testchart/values.yaml: image: testchart/testimage:{tag}" in out
 

